File: insertion.c
//Write a program to perform insertion of an element/node in a given list.

#include <stdio.h>
#include <stdlib.h>

// Node structure
struct Node {
    int data;
    struct Node* next;
};

struct Node* head = NULL;

// Insert at the beginning
void insertAtBeginning(int value) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = value;
    newNode->next = head;
    head = newNode;
}

// Insert at the end
void insertAtEnd(int value) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = value;
    newNode->next = NULL;
    
    if (head == NULL) {
        head = newNode;
        return;
    }

    struct Node* temp = head;
    while (temp->next != NULL)
        temp = temp->next;
    temp->next = newNode;
}

// Insert at a specific position (0-based index)
void insertAtPosition(int value, int pos) {
    if (pos == 0) {
        insertAtBeginning(value);
        return;
    }
    int i;

    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = value;

    struct Node* temp = head;
    for (i = 0; i < pos - 1 && temp != NULL; i++) {
        temp = temp->next;
    }

    if (temp == NULL) {
        printf("Position out of bounds!\n");
        return;
    }

    newNode->next = temp->next;
    temp->next = newNode;
}

// Display the list
void display() {
    struct Node* temp = head;
    printf("List: ");
    while (temp != NULL) {
        printf("%d -> ", temp->data);
        temp = temp->next;
    }
    printf("NULL\n");
}

int main() {
    int choice, value, position;

    while (1) {
        printf("\n1. Insert at beginning\n2. Insert at end\n3. Insert at position\n4. Display\n5. Exit\n");
        printf("Enter your choice: ");
        scanf("%d", &choice);

        switch (choice) {
            case 1:
                printf("Enter value: ");
                scanf("%d", &value);
                insertAtBeginning(value);
                break;
            case 2:
                printf("Enter value: ");
                scanf("%d", &value);
                insertAtEnd(value);
                break;
            case 3:
                printf("Enter value and position: ");
                scanf("%d %d", &value, &position);
                insertAtPosition(value, position);
                break;
            case 4:
                display();
                break;
            case 5:
                exit(0);
            default:
                printf("Invalid choice!\n");
        }
    }

    return 0;
}

File: insertion.py
#Write a program to perform insertion of an element/node in a given list.

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def insert_at_beginning(self, value):
        new_node = Node(value)
        new_node.next = self.head
        self.head = new_node

    def insert_at_end(self, value):
        new_node = Node(value)
        if not self.head:
            self.head = new_node
            return
        temp = self.head
        while temp.next:
            temp = temp.next
        temp.next = new_node

    def insert_at_position(self, value, pos):
        if pos == 0:
            self.insert_at_beginning(value)
            return
        new_node = Node(value)
        temp = self.head
        for _ in range(pos - 1):
            if not temp:
                print("Position out of bounds!")
                return
            temp = temp.next
        if not temp:
            print("Position out of bounds!")
            return
        new_node.next = temp.next
        temp.next = new_node

    def display(self):
        temp = self.head
        print("List:", end=" ")
        while temp:
            print(f"{temp.data} -> ", end="")
            temp = temp.next
        print("NULL")

# Example usage
ll = LinkedList()
while True:
    print("\n1. Insert at beginning\n2. Insert at end\n3. Insert at position\n4. Display\n5. Exit")
    choice = int(input("Enter your choice: "))
    
    if choice == 1:
        val = int(input("Enter value: "))
        ll.insert_at_beginning(val)
    elif choice == 2:
        val = int(input("Enter value: "))
        ll.insert_at_end(val)
    elif choice == 3:
        val = int(input("Enter value: "))
        pos = int(input("Enter position (0-based index): "))
        ll.insert_at_position(val, pos)
    elif choice == 4:
        ll.display()
    elif choice == 5:
        break
    else:
        print("Invalid choice!")
